### DISCUSSION TITLE
CLI plugin using processWithSession which is passed the session as AbstractSession not using token in one command

### DISCUSSION THREAD

**Discussion by davidkjackson54** [zowe/imperative#1009] [Q&A] [answered]
In my zowe cli plugin, I have some 30-40 commands defined. All follow the same  structure and layout.
In my System Profile, I have tokenType and tokenValue present.
```
"properties": {
                        "basePath": "/zmfrest",
                        "port": 9992,
                        "protocol": "https",
                        "zosmfProfile": "zosmf",
                        "tokenType": "JSESSIONID",
                        "tokenValue": "6033AA612A70926297D26E19840EC62A"
```
I am finding that in one command that happens to be called "logon" in a group called 'session" that for whatever reason, the session: AbstractSession is being passed into my handler as type basic and is using user and password -when it should be using tokenValue.
Here is the screen print of params:IHandlerParams.arguments - notice that user and password are present ..
![image](https://github.com/zowe/imperative/assets/46171060/00141d99-b216-4ea7-8490-dc55ebec7cd5)
![image](https://github.com/zowe/imperative/assets/46171060/91b6a9bf-fe98-4ff5-93af-3a3dadd17838)

and this is the session:AbstractSession that is being passed in:
![image](https://github.com/zowe/imperative/assets/46171060/c4f79e3a-cfe3-4a24-b4bd-6a5aa131814f)

notice that the session is a 'basic' session using user and password and not the token. I can see no reason why this incoming session is uinsg user and password. 

I am finding that other commands are indeed working correctly and using tokenType and tokenValue 
as in this example of another command:
![image](https://github.com/zowe/imperative/assets/46171060/b3b21b00-11b1-4adb-b1a0-5371130977ae)
![image](https://github.com/zowe/imperative/assets/46171060/213aa46f-92c0-4bc7-bce1-ad7b8b992ed5)
and its session:AbstractSession
![image](https://github.com/zowe/imperative/assets/46171060/b505765f-d254-4304-a4d6-450bcfbca408)

The definitions file:
![image](https://github.com/zowe/imperative/assets/46171060/65ed1001-a0b8-4af0-b26f-66eb6144f101)

it is the logon definitions that are incorrectly using the user/password whilst the logoff command is correctly using the token.

I can see no obvious reason why this is happening and unable to debug this further.

**Comment by t1m0thyj** [comment]
@davidkjackson54 Is it possible to show the code for `ZmfLogonDefinition` and `ZmfLogoffDefinition` (with any private info redacted)?

Since one behaves as expected and the other doesn't, I'd be curious to compare these two command definitions.

**Comment by davidkjackson54** [comment, marked as answer]
@t1m0thyj 

I have resolved this after walking the code in Imperative  ConnectionPropsForSessCfg. It gave me some pointers as to what might be going on.  It reminded me that some 2 years or so back, that I had specifically put code in to force my LOGON command to NOT use a token if it existed. as it had most likely expired and the purpose of the logon command was to acquire a new token.

My code:
```
if (params.definition.name != "zmflogon" && params.arguments.tokenValue) {
      // if a call other than a zmflogon and a token exists then use it
      // if a token is available, then remove all reference to userid and password to stop zowe creating a type "Basic" session
      delete params.arguments.user;
      delete params.arguments.u;
      delete params.arguments.password;
      delete params.arguments.pass;
      delete params.arguments.pw;
    }

    const session = await ZmfSession.createSessCfgFromArgs(params.arguments); 
```
